File: core/server/project/shine/src/main/java/com/home/shine/dataEx/TimeExpression.java
package com.home.shine.dataEx;

import com.home.shine.control.DateControl;
import com.home.shine.ctrl.Ctrl;
import com.home.shine.support.CronExpression;
import com.home.shine.timer.DefaultTimeEntity;
import com.home.shine.timer.ITimeEntity;
import com.home.shine.utils.TimeUtils;

import java.text.ParseException;

/** 时间表达式 */
public class TimeExpression
{
	/** 空 */
	private static final int None=0;
	/** 秒倒计时 */
	private static final int SecondTimeOut=1;
	/** cron表达式 */
	private static final int Cron=9;
	
	/** 类型 */
	private int _type;
	/** 参数 */
	private int _arg;
	/** cron表达式 */
	private CronExpression _cron;
	
	public TimeExpression(String str)
	{
		if(str==null || str.isEmpty())
		{
			_type=None;
			return;
		}
		
		int index=str.indexOf(":");
		
		if(index!=-1)
		{
			_type=Integer.parseInt(str.substring(0,index));
			_arg=Integer.parseInt(str.substring(index+1,str.length()));
		}
		else
		{
			_type=Cron;
			_cron=TimeUtils.createCronExpression(str);
		}
	}
	
	/** 是否为空 */
	public boolean isEmpty()
	{
		return _type==None;
	}

	/** 获取下个时间(ms)(如为空返回-1L) */
	public long getNextTime(long from)
	{
		switch(_type)
		{
			case None:
			{
				return -1L;
			}
			case SecondTimeOut:
			{
				return from+(_arg*1000);
			}
			case Cron:
			{
				return TimeUtils.getNextCronTime(_cron,from);
			}
			default:
			{
				Ctrl.throwError("不支持的时间表达式类型",_type);
			}
			break;
		}

		return from;
	}
	
	/** 获取上个时间(ms)(如为空返回-1L) */
	public long getPrevTime(long from)
	{
		switch(_type)
		{
			case None:
			{
				return -1L;
			}
			case SecondTimeOut:
			{
				return from-(_arg*1000);
			}
			case Cron:
			{
				return TimeUtils.getPrevCronTime(_cron,from);
			}
			default:
			{
				Ctrl.throwError("不支持的时间表达式类型",_type);
			}
			break;
		}
		
		return from;
	}
	
	/** 获取下个时间(ms) */
	public static long getNextTimeS(String str,long time)
	{
		return new TimeExpression(str).getNextTime(time);
	}

	/** 获取下个时间(ms) */
	public long getNextTime(ITimeEntity entity)
	{
		return getNextTime(entity.getTimeMillis());
	}
	
	/** 获取下个时间(ms) */
	public long getNextTime()
	{
		return getNextTime(DateControl.getTimeMillis());
	}
	
	/** 获取上个时间(ms) */
	public long getPrevTime()
	{
		return getPrevTime(DateControl.getTimeMillis());
	}
	
	/** 获取上个时间(ms) */
	public long getPrevTime(ITimeEntity entity)
	{
		return getPrevTime(entity.getTimeMillis());
	}
	
	/** 获取下个时间(ms) */
	public static long getNextTimeS(String str)
	{
		return new TimeExpression(str).getNextTime();
	}
}

File: core/clientLaya/game/src/shine/dataEx/TimeExpression.ts
namespace Shine
{
	export class TimeExpression
	{
		/** 空 */
		private None:number=0;
		/** 秒倒计时 */
		private SecondTimeOut:number=1;
		/** cron表达式 */
		private Cron:number=9;
	
		/** 类型 */
		private _type:number;
		/** 参数 */
		private _arg:number;
		// /** cron表达式 */
		// private _cron:CronExpression;
	
		constructor(str:string)
		{
			if(str==null || str.length==0)
			{
				this._type=this.None;
				return;
			}
	
			var index:number=str.indexOf(":");
	
			if(index!=-1)
			{
				this._type=parseInt(str.slice(0,index));
				this._arg=parseInt(str.slice(index+1,str.length));
			}
			else
			{
				this._type=this.Cron;
	
				try
				{
					// this._cron=new CronExpression(str);
				}
				catch(err)
				{
					Ctrl.throwError("cron表达式错误",err);
				}
			}
		}
	
		/** 是否为空 */
		public isEmpty():boolean
		{
			return this._type==this.None;
		}
	
		/** 获取下个时间(秒)(如为空返回-1L) */
		public getNextTime(from:number=DateControl.getTimeMillis()):number
		{
			switch(this._type)
			{
				case this.None:
				{
					return -1;
				}
				case this.SecondTimeOut:
				{
					return from+this._arg;
				}
				case this.Cron:
				{
					// return TimeUtils.getNextCronTime(this._cron,from);
					return 0;
				}
				default:
				{
					Ctrl.throwError("不支持的时间表达式类型",this._type);
				}
					break;
			}
	
			return from;
		}

		/** 获取上个时间(ms)(如为空返回-1L) */
		public getPrevTime(from:number=DateControl.getTimeMillis()):number
		{
			switch(this._type)
			{
				case this.None:
				{
					return -1;
				}
				case this.SecondTimeOut:
				{
					return from-(this._arg*1000);
				}
				case this.Cron:
				{
					// return TimeUtils.getPrevCronTime(_cron,from);
					return 0;
				}
				default:
				{
					Ctrl.throwError("不支持的时间表达式类型",this._type);
				}
				break;
			}
			
			return from;
		}

		/** 获取下个时间 */
		public static getNextTimeS(str:string ,time:number=DateControl.getTimeMillis()):number
		{
			// return new TimeExpression(str).getNextTime(time);
			return 0;
		}
	}
}